aten/src/ATen/cpu/vec256/vec256_complex_double.h

```diff
@@ -0,0 +1,342 @@
+                                        base + step);
+  }
+  static Vec256<std::complex<double>> set(const Vec256<std::complex<double>>& a, const Vec256<std::complex<double>>& b,
+                            int64_t count = size()) {
+    switch (count) {
+      case 0:
+        return a;
+      case 1:
+        return blend<1>(a, b);
+    }
+    return b;
+  }
+  static Vec256<std::complex<double>> loadu(const void* ptr, int64_t count = size()) {
+    if (count == size())
+      return _mm256_loadu_pd(reinterpret_cast<const double*>(ptr));
+
+    __at_align32__ double tmp_values[2*size()];
+    std::memcpy(
+        tmp_values,
+        reinterpret_cast<const double*>(ptr),
+        count * sizeof(std::complex<double>));
+    return _mm256_load_pd(tmp_values);
+  }
+  void store(void* ptr, int count = size()) const {
+    if (count == size()) {
+      _mm256_storeu_pd(reinterpret_cast<double*>(ptr), values);
+    } else if (count > 0) {
+      double tmp_values[2*size()];
+      _mm256_storeu_pd(reinterpret_cast<double*>(tmp_values), values);
+      std::memcpy(ptr, tmp_values, count * sizeof(std::complex<double>));
+    }
+  }
+  const std::complex<double>& operator[](int idx) const  = delete;
+  std::complex<double>& operator[](int idx) = delete;
+  Vec256<std::complex<double>> map(std::complex<double> (*f)(const std::complex<double> &)) const {
+    __at_align32__ std::complex<double> tmp[size()];
+    store(tmp);
+    for (int i = 0; i < size(); i++) {
+      tmp[i] = f(tmp[i]);
+    }
+    return loadu(tmp);
+  }
+  __m256d abs_2_() const {
+    auto val_2 = _mm256_mul_pd(values, values);     // a*a     b*b
+    return _mm256_hadd_pd(val_2, val_2);            // a*a+b*b a*a+b*b
+  }
+  __m256d abs_() const {
+    auto abs = _mm256_sqrt_pd(abs_2_());            // abs     abs
+    auto mask = _mm256_setr_pd(1.0, 0.0, 1.0, 0.0);
+    return _mm256_mul_pd(abs, mask);                // abs     0
+  }
+  Vec256<std::complex<double>> abs() const {
+   return abs_();     // abs     0
+  }
+  __m256d angle_() const {
+    auto b_a = _mm256_permute_pd(values, 0x05);      //b        a
+    auto angle = Sleef_atan2d4_u10(values, b_a);     //-angle   angle
+    auto mask = _mm256_setr_pd(0.0, 1.0, 0.0, 1.0);
+    return _mm256_mul_pd(angle, mask);               //0        angle
+  }
+  Vec256<std::complex<double>> angle() const {
+    return _mm256_permute_pd(angle_(), 0x05);        // angle     0
+  }
+  __m256d real_() const {
+    auto mask = _mm256_setr_pd(1.0, 0.0, 1.0, 0.0);
+    return _mm256_mul_pd(values, mask);
+  }
+  Vec256<std::complex<double>> real() const {
+    return real_();
+  }
+  __m256d imag_() const {
+    auto mask = _mm256_setr_pd(0.0, 1.0, 0.0, 1.0);
+    return _mm256_mul_pd(values, mask);
+    }
+  Vec256<std::complex<double>> imag() const {
+    return _mm256_permute_pd(imag_(), 0x05);        //b        a
+  }
+  Vec256<std::complex<double>> acos() const {
+    return map(std::acos);
```

Comment:
Yeah, I actually couldn't figure that out. I read that template classes don't support inheritance? Am I crazy?

Reply:
Non-virtual template methods inherit just fine. https://godbolt.org/z/Gt1QfK

Reply:
I looked into it. Vec256 is a template class. While inheritance is perfectly valid 
eg `class Vec256z : public Vec256<std:complex<double>>{}` it requires that I define a new class `Vec256z`. The rest of the code is creating template objects using `Vec256<scalar_t>`.  In your example code you needed to create a new class name `B`, whereas I have to use `Vec256`.

You may notice that `Vec256<int64_t>` only defines several methods of `Vec256<T>`. I think this is because very few math kernels use integer specific function calls.